core/src/main/java/org/infinispan/util/concurrent/CommandAckCollector.java

```diff
@@ -85,17 +87,19 @@ public void inject(
       if (backupOwners.isEmpty()) {
          return new PrimaryOwnerOnlyCollector<>();
       }
-      SingleKeyCollector collector = new SingleKeyCollector(id, backupOwners, topologyId);
+      SingleKeyCollector<T> collector = new SingleKeyCollector<>(id, backupOwners, topologyId);
       BaseAckTarget prev = collectorMap.put(id, collector);
-      assert prev == null : prev.toString();
+      //is it possible the have a previous collector when the topology changes after the first collector is created
+      //in that case, the previous collector must have a lower topology id
+      assert prev == null || prev.topologyId < topologyId : format("replaced old collector '%s' by '%s'", prev, collector);
```

Comment:
So, the only "fix" is an additional assert to verify an assumption. How does this "fix" things ? What if the user runs with assertions disabled ?

Reply:
the code is correct but the assertion wasn't. the test was failing because of the assertion. 
if you disable the assertions, the test is fine (and the code as well)